test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/ViewComponents/DefaultViewComponentHelperTest.cs

```diff
@@ -0,0 +1,126 @@
+using Microsoft.Extensions.WebEncoders.Testing;
+using Moq;
+using Xunit;
+
+namespace Microsoft.AspNetCore.Mvc.ViewComponents
+{
+    public class DefaultViewComponentHelperTest
+    {
+        [Fact]
+        public void GetArgumentDictionary_SupportsAnonymouslyTypedArguments()
+        {
+            // Arrange
+            var descriptorCollectionProvider = Mock.Of<IViewComponentDescriptorCollectionProvider>();
+            var selector = Mock.Of<IViewComponentSelector>();
+            var invokerFactory = Mock.Of<IViewComponentInvokerFactory>();
+            var viewBufferScope = Mock.Of<IViewBufferScope>();
+
+            var helper = new DefaultViewComponentHelper(
+                descriptorCollectionProvider,
+                new HtmlTestEncoder(),
+                selector,
+                invokerFactory,
+                viewBufferScope);
+
+            var descriptor = CreateDescriptorForType(typeof(TestViewComponent));
+
+            // Act
+            var argumentDictionary = helper.GetArgumentDictionary(descriptor, new { a = 0 });
+
+            // Assert
+            Assert.Equal(1, argumentDictionary.Count);
+            Assert.Collection(argumentDictionary,
+                item =>
+                {
+                    Assert.Equal("a", item.Key);
+                    Assert.IsType(typeof(int), item.Value);
+                    Assert.Equal(0, item.Value);
+                });
+        }
+
+        [Fact]
+        public void GetArgumentDictionary_SingleParameter_DoesNotNeedAnonymouslyTypedArguments()
+        {
+            // Arrange
+            var descriptorCollectionProvider = Mock.Of<IViewComponentDescriptorCollectionProvider>();
+            var selector = Mock.Of<IViewComponentSelector>();
+            var invokerFactory = Mock.Of<IViewComponentInvokerFactory>();
+            var viewBufferScope = Mock.Of<IViewBufferScope>();
+
+            var helper = new DefaultViewComponentHelper(
+                descriptorCollectionProvider,
+                new HtmlTestEncoder(),
+                selector,
+                invokerFactory,
+                viewBufferScope);
+
+            var descriptor = CreateDescriptorForType(typeof(TestViewComponent));
+
+            // Act
+            var argumentDictionary = helper.GetArgumentDictionary(descriptor, 0);
+
+            // Assert
+            Assert.Equal(1, argumentDictionary.Count);
+            Assert.Collection(argumentDictionary,
+                item =>
+                {
+                    Assert.Equal("a", item.Key);
+                    Assert.IsType(typeof(int), item.Value);
+                    Assert.Equal(0, item.Value);
+                });
+        }
+
+        private ViewComponentDescriptor CreateDescriptorForType(Type componentType)
+        {
+            var provider = CreateProvider(componentType);
+            return provider.GetViewComponents().First();
+        }
+
+        private class TestViewComponent
```

Comment:
Could we have ViewComponents with multiple params, with `object` param type and one with `IDictionary<string, object>`. The last two are special cases in the `IsAssignableFrom` \ `PropertyHelper` scenarios. The first one covers the alternate code path in `GetArgumentDictionary`